beep/structure/base.py

```diff
@@ -1385,3 +1412,62 @@ def get_max_paused_over_threshold(group, paused_threshold=3600):
     else:
         max_paused_duration = 0
     return max_paused_duration
+
+
+def get_CV_segment_from_charge(charge):
+    """
+    Extracts the constant voltage segment from charge. Works for both CCCV or
+    CC steps followed by a CV step.
+
+    Args:
+        charge (pd.DataFrame): charge dataframe for a single cycle
+
+    Returns:
+        (pd.DataFrame): dataframe containing the CV segment
+
+    """
+    # Compute dI and dV
+    dI = np.diff(charge.current)
+    dV = np.diff(charge.voltage)
+    dt = np.diff(charge.test_time)
+
+    # Find the first index where dt>1 and abs(dV/dt)<tol and abs(dI/dt)>tol
+    i = 0
+    while i < len(dV) and not (dt[i] > 1 and abs(dV[i]/dt[i]) < 5.e-5 and abs(dI[i]/dt[i]) > 1.e-4):
```

Comment:
Also, is there a better way of evaluating this condition or narrowing down the number of points this is evaluated on? There are two nested loops here that give me pause. Can we try this code out on a full size file to check the compute time?

Reply:
I used the while loop to reduce the number of evaluated points. When the beginning of CV is reached, the loop breaks and no more points are evaluated. Would it be faster to avoid the while loop but evaluate every point? I can measure some timings.

I think the nesting might be necessary. For every cycle, you have loop over the charge points and find the start of CV.

Reply:
Evaluating get_CV_segment_from_charge() on ~900 cycles of raw data takes ~9 seconds on my workstation. Most of that time is filtering each cycle for charge, but replacing the inner while loop with vectorized numpy calls saves ~1sec. Should I make this change?

Reply:
I guess if its only a small improvement we can leave as is for now, but it might be nice to include the numpy call as a comment in case we need a speed improvement.

Reply:
The numpy approach had problems with boundary cases (empty arrays and data with no CV), so let's stick with the loop for now.